okhttp/src/jvmMain/kotlin/okhttp3/Dispatcher.kt

```diff
@@ -164,12 +196,27 @@ class Dispatcher() {
     val executableCalls = mutableListOf<AsyncCall>()
     val isRunning: Boolean
     synchronized(this) {
-      val i = readyAsyncCalls.iterator()
-      while (i.hasNext()) {
-        val asyncCall = i.next()
+      while (readyAsyncCalls.isNotEmpty()) {
+        val filteredCalls = readyAsyncCalls.mapNotNull {
+          if (it.callsPerHost.get() < maxRequestsPerHost) it.call else null
+        }
+
+        if (filteredCalls.isEmpty()) break // Max host capacity for all requests.
+
+        val nextCall = asyncCallDispatcher.nextCall(filteredCalls)
```

Comment:
I worry about putting user logic in this hot path.  Potentially that needs to process through large list of items, on every pass through promoteAndExecute.  I wonder whether a more constrained version without the same risk, is to return a priority for each item, with a default returning 0 for all.  These have a priority queue for readyAsyncCalls.